File: app_express/src/repositories/bonRepository.js
const { applyDateRange, assertNoError } = require('./base');
const { PAYMENT_STATUS, applyFactoryDeductionPresets } = require('../services/calculations');
const { notifyChange } = require('../services/realtimeService');

function serializeBon(body, calculated, imageUrl) {
  const data = {
    ticket_number: body.ticket_number || null,
    bon_date: body.bon_date,
    plate_number: String(body.plate_number || '').replace(/\s+/g, '').toUpperCase(),
    driver_name: body.driver_name ? String(body.driver_name).trim().toUpperCase() : null,
    relation_name: body.relation_agent_id ? null : (body.relation_name ? String(body.relation_name).trim().toUpperCase() : null),
    relation_agent_id: body.relation_agent_id || null,
    factory_id: body.factory_id || null,
    factory_spsi_type_id: body.factory_spsi_type_id || null,
    spsi_type_name: body.spsi_type_name ? String(body.spsi_type_name).trim().toUpperCase() : null,
    spsi_calculation_mode: calculated.spsi_calculation_mode,
    spsi_rate: calculated.spsi_rate,
    spsi_amount: calculated.spsi_amount,
    fruit_origin: body.fruit_origin ? String(body.fruit_origin).trim().toUpperCase() : null,
    chat_id: body.chat_id || null,
    message_id: body.message_id || null,
    notes: body.notes || null,
    netto_1: calculated.netto_1,
    netto_2: calculated.netto_2,
    price: calculated.price,
    dp: calculated.dp,
    biaya_bongkar: calculated.biaya_bongkar,
    bp_colt: calculated.bp_colt,
    pph: calculated.pph,
    uang_minum: calculated.uang_minum,
    total: calculated.total,
    status: body.status || PAYMENT_STATUS.BELUM_DIBAYAR
  };

  if (imageUrl) data.image_url = imageUrl;
  return data;
}

async function listBons(supabase, filters = {}) {
  let query = supabase.from('bons').select('*, bon_deductions(*), factories(name), relation_agents(name)');
  query = applyDateRange(query, 'bon_date', filters.start, filters.end);
  if (filters.status) query = query.eq('status', filters.status);
  if (filters.q) {
    const q = String(filters.q).trim();
    query = query.or(`driver_name.ilike.%${q}%,plate_number.ilike.%${q}%,relation_name.ilike.%${q}%`);
  }
  if (filters.factory_id) query = query.eq('factory_id', filters.factory_id);
  return assertNoError(await query.order('created_at', { ascending: false }));
}

async function getBon(supabase, id) {
  return assertNoError(
    await supabase.from('bons').select('*, bon_deductions(*), factories(name), relation_agents(name)').eq('id', id).single()
  );
}

async function getLatestPrice(supabase) {
  const { data, error } = await supabase
    .from('bons')
    .select('price')
    .order('created_at', { ascending: false })
    .limit(1)
    .maybeSingle();
  if (error) throw error;
  return data?.price || 0;
}

async function createBon(supabase, data, deductions = []) {
  const created = assertNoError(
    await supabase.from('bons').insert(data).select().single()
  );

  const deductionsToSave = applyFactoryDeductionPresets(data.factory_id, deductions);
  if (deductionsToSave.length > 0) {
    assertNoError(
      await supabase.from('bon_deductions').insert(
        deductionsToSave.map((item) => ({
          bon_id: created.id,
          label: item.label,
          amount: item.amount
        }))
      )
    );
  }

  notifyChange('bons', 'INSERT', created);
  return getBon(supabase, created.id);
}

async function updateBon(supabase, id, data, deductions = [], skipStatusCheck = false) {
  if (!skipStatusCheck) {
    const current = assertNoError(await supabase.from('bons').select('status').eq('id', id).single());
    if (current.status === PAYMENT_STATUS.LUNAS) {
      throw new Error('Bon sudah lunas, tidak dapat diedit.');
    }
  }

  assertNoError(await supabase.from('bons').update(data).eq('id', id));
  assertNoError(await supabase.from('bon_deductions').delete().eq('bon_id', id));
  if (deductions.length > 0) {
    assertNoError(
      await supabase.from('bon_deductions').insert(
        deductions.map((item) => ({ bon_id: id, label: item.label, amount: item.amount }))
      )
    );
  }
  const updated = await getBon(supabase, id);
  notifyChange('bons', 'UPDATE', updated);
  return updated;
}

async function deleteBon(supabase, id) {
  const current = assertNoError(await supabase.from('bons').select('status, image_url, plate_number, ticket_number').eq('id', id).single());
  if (current.status !== PAYMENT_STATUS.BELUM_DIBAYAR) {
    throw new Error('Bon sudah diproses (Tertagih/Lunas), tidak dapat dihapus.');
  }
  // Hapus file dari bucket jika ada
  if (current.image_url) {
    const pathMatch = current.image_url.match(/\/object\/public\/[^/]+\/(.+)$/);
    if (pathMatch) {
      const filePath = pathMatch[1];
      await supabase.storage.from('receipts').remove([filePath]).catch(() => {});
    }
  }
  assertNoError(await supabase.from('bons').delete().eq('id', id));
  notifyChange('bons', 'DELETE', null, current);
}

async function getRelatedRecords(supabase, bonId) {
  const notaItems = assertNoError(
    await supabase.from('nota_items').select('notas(*)').eq('bon_id', bonId)
  );
  const notas = (notaItems || []).map((item) => item.notas).filter(Boolean);
  if (notas.length === 0) return { notas: [], payments: [] };

  const notaIds = notas.map((nota) => nota.id);
  const payments = assertNoError(
    await supabase.from('payments').select('*, notas(invoice_number)').in('invoice_id', notaIds)
  );
  return { notas, payments };
}

module.exports = {
  createBon,
  deleteBon,
  getBon,
  getLatestPrice,
  getRelatedRecords,
  listBons,
  serializeBon,
  updateBon
};

File: lib/repositories/bon_repository.dart
import 'dart:typed_data';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../../models/bon_model.dart';
import '../../models/nota_model.dart';
import '../../models/payment_model.dart';

class BonRepository {
  final SupabaseClient _client;

  BonRepository(this._client);

  Future<List<BonModel>> getBons({
    DateTime? startDate,
    DateTime? endDate,
    String? driverQuery,
    String? factoryId,
  }) async {
    var query = _client
        .from('bons')
        .select('*, bon_deductions(*), factories(name), relation_agents(name)');

    if (startDate != null) {
      query = query.gte('bon_date', startDate.toIso8601String().split('T')[0]);
    }
    if (endDate != null) {
      final nextDay = DateTime.utc(endDate.year, endDate.month, endDate.day + 1);
      query = query.lt('bon_date', nextDay.toIso8601String().split('T')[0]);
    }
    if (driverQuery != null && driverQuery.isNotEmpty) {
      query = query.or(
        'driver_name.ilike.%$driverQuery%,plate_number.ilike.%$driverQuery%,relation_name.ilike.%$driverQuery%',
      );
    }
    if (factoryId != null && factoryId.isNotEmpty) {
      query = query.eq('factory_id', factoryId);
    }

    final response = await query.order('created_at', ascending: false);
    return (response as List).map((e) => BonModel.fromJson(e)).toList();
  }

  Future<BonModel> createBon(
    BonModel bon,
    Uint8List? imageBytes,
    String? fileName, {
    String? existingImageUrl,
  }) async {
    String? imageUrl = existingImageUrl;
    if (imageUrl == null && imageBytes != null && fileName != null) {
      final path = 'bons/${DateTime.now().millisecondsSinceEpoch}_$fileName';
      await _client.storage.from('receipts').uploadBinary(path, imageBytes);
      imageUrl = _client.storage.from('receipts').getPublicUrl(path);
    }

    final bonData = bon.toJson();
    if (imageUrl != null) {
      bonData['image_url'] = imageUrl;
    }
    bonData.remove('id');

    final response = await _client
        .from('bons')
        .insert(bonData)
        .select()
        .single();
    final createdBon = BonModel.fromJson(response);

    // Save deductions if any
    if (bon.deductions.isNotEmpty) {
      final deductionsData = bon.deductions
          .map(
            (d) => {
              'bon_id': createdBon.id,
              'label': d.label,
              'amount': d.amount,
            },
          )
          .toList();
      await _client.from('bon_deductions').insert(deductionsData);
    }

    // Refresh to get deductions back
    return (await _client
        .from('bons')
        .select('*, bon_deductions(*), factories(name), relation_agents(name)')
        .eq('id', createdBon.id)
        .single()
        .then((v) => BonModel.fromJson(v)));
  }

  Future<void> quickUpdateBon(String id, Map<String, dynamic> changes) async {
    final current = await _client
        .from('bons')
        .select('*, bon_deductions(amount)')
        .eq('id', id)
        .single();
    if (current['status'] == 'LUNAS') {
      throw Exception('Bon sudah lunas, tidak dapat diedit.');
    }

    final merged = Map<String, dynamic>.from(current);
    merged.addAll(changes);

    final netto1 = (merged['netto_1'] as num?)?.toInt() ?? 0;
    final netto2 = (merged['netto_2'] as num?)?.toInt() ?? 0;
    final price = (merged['price'] as num?)?.toInt() ?? 0;
    final dp = (merged['dp'] as num?)?.toInt() ?? 0;
    final bpColt = (merged['bp_colt'] as num?)?.toInt() ?? 0;
    final pph = (merged['pph'] as num?)?.toInt() ?? 0;
    final uangMinum = (merged['uang_minum'] as num?)?.toInt() ?? 0;
    final deductionTotal = ((merged['bon_deductions'] as List?) ?? [])
        .fold<int>(0, (sum, d) => sum + ((d['amount'] as num?)?.toInt() ?? 0));
    final spsiMode = (merged['spsi_calculation_mode'] as String?) ?? 'PER_KG';
    final spsiRate =
        (merged['spsi_rate'] as num?)?.toInt() ??
        (merged['biaya_bongkar'] as num?)?.toInt() ??
        12;
    final subtotal = price * netto2;
    final totalBiayaBongkar = spsiMode == 'FIX' ? spsiRate : spsiRate * netto1;
    final total =
        subtotal -
        dp -
        totalBiayaBongkar -
        bpColt -
        pph -
        uangMinum -
        deductionTotal;

    final data = <String, dynamic>{
      'netto_2': netto2,
      'price': price,
      'bp_colt': bpColt,
      'pph': pph,
      'uang_minum': uangMinum,
      'spsi_amount': totalBiayaBongkar,
      'total': total,
    };

    await _client.from('bons').update(data).eq('id', id);
  }

  Future<void> updateRaw(String id, Map<String, dynamic> data) async {
    await _client.from('bons').update(data).eq('id', id);
  }

  Future<void> updateBon(BonModel bon) async {
    // 1. Check current status in DB
    final current = await _client
        .from('bons')
        .select('status')
        .eq('id', bon.id)
        .single();
    if (current['status'] == 'LUNAS') {
      throw Exception('Bon sudah lunas, tidak dapat diedit.');
    }

    final data = bon.toJson();
    data.remove('id');
    await _client.from('bons').update(data).eq('id', bon.id);

    // Update deductions: Delete and Re-insert for simplicity
    await _client.from('bon_deductions').delete().eq('bon_id', bon.id);
    if (bon.deductions.isNotEmpty) {
      final deductionsData = bon.deductions
          .map((d) => {'bon_id': bon.id, 'label': d.label, 'amount': d.amount})
          .toList();
      await _client.from('bon_deductions').insert(deductionsData);
    }
  }

  Future<void> deleteBon(String id) async {
    // 1. Check current status in DB
    final current = await _client
        .from('bons')
        .select('status, image_url')
        .eq('id', id)
        .single();
    if (current['status'] != 'BELUM_DIBAYAR') {
      throw Exception('Bon sudah dibuat nota atau lunas, tidak dapat dihapus.');
    }

    // 2. Hapus file dari bucket jika ada
    final imageUrl = current['image_url'] as String?;
    if (imageUrl != null && imageUrl.isNotEmpty) {
      final pathMatch = RegExp(
        r'/object/public/[^/]+/(.+)$',
      ).firstMatch(imageUrl);
      if (pathMatch != null) {
        final filePath = pathMatch.group(1)!;
        await _client.storage
            .from('receipts')
            .remove([filePath])
            .catchError((_) {});
      }
    }

    await _client.from('bons').delete().eq('id', id);
  }

  Future<Map<String, List<dynamic>>> getRelatedRecords(String bonId) async {
    // 1. Get Notas via nota_items
    final notaItems = await _client
        .from('nota_items')
        .select('notas(*)')
        .eq('bon_id', bonId);

    final notasData = notaItems.map((e) => e['notas']).toList();

    if (notasData.isEmpty) {
      return {'notas': [], 'payments': []};
    }

    final notaIds = notasData.map((i) => i['id'] as String).toList();

    // 2. Get Payments for those notas
    final paymentsResponse = await _client
        .from('payments')
        .select('*, notas(invoice_number)')
        .filter('invoice_id', 'in', notaIds);

    return {
      'notas': notasData.map((e) => NotaModel.fromJson(e)).toList(),
      'payments': (paymentsResponse as List)
          .map((e) => PaymentModel.fromJson(e))
          .toList(),
    };
  }

  Future<double> getLatestPrice() async {
    try {
      final response = await _client
          .from('bons')
          .select('price')
          .order('created_at', ascending: false)
          .limit(1)
          .maybeSingle();

      if (response != null && response['price'] != null) {
        return (response['price'] as num).toDouble();
      }
    } catch (_) {
      // Return 0 if fails
    }
    return 0.0;
  }
}
